### ISSUE TITLE
CSS optimization techniques?

### ISSUE THREAD

**Issue by LukaszJaro** [digitoimistodude/air-light#86]
Hi, 

just learning about this starter theme, lots of great things going on here! I'm curious if there is anything radical being done to optimize the CSS like in this starter theme which loads the CSS only when the component is in view:

https://github.com/wprig/docs/blob/master/documentation/css.md#progressive-loading-of-css

**Comment by ronilaukkarinen** [comment]
I actually wrote a blog post about his (sadly [in Finnish](https://www.dude.fi/verkkosivujen-latausnopeus)). I may write about this also in English on [rolle.design](https://rolle.design). We think it's better to have styles loaded up once than multiple files/http requests because otherwise new file is loaded with every new page switch, right? This is a matter of approach. Also we want to keep the main CSS files as small as possible so we are talking about saves like couple of dozens kilobytes per request. Doesn't sound worth the trouble.

But, to be honest these techniques are not that familiar to us. I have to read about wprig on better time. If you have ideas how to implement this on Air-light, please do tell (I'll leave the issue open for now) or send a PR for review. Thanks for taking the time thinking about making things better! 👍

**Comment by LukaszJaro** [comment]
Thanks! I translated your blog post with google although it doesn't work perfectly :) -I get the point of what your trying to say.

> It makes much more sense for the visitor to get the entire site code with them on the first download, so that the subpages load faster when everything is already loaded into the user’s browser cache at once.

Feel free to look at wprig, the original owner of the repo is actually from Norway :)

What I like about your starter theme is it keeps more to WP way of things, and comes with very practical and useful things. I don't think it's necessary to implement it how wprig is doing it..I'll see how it goes building a site with the air-light to get better idea of css loading, example if @import is making another http request or build process writes @imports to css file:

> Files prefixed by a _ (eg _blocks.css) are partials, compiled into their parent files using @import rules. PostCSS imports the contents of these partials during processing to become part of the parent file.

**Comment by ronilaukkarinen** [comment]
> Feel free to look at wprig, the original owner of the repo is actually from Norway :)

Norway is a neighbour of a neighbour of Finland, our direct neighbours are Sweden and Russia :) Nice little detail anyway!

I just checked more about wprig. Looks interesting but quite a different approach and I don't see how it would fit to our project. Again, PR/ideas are welcome on how to impelement or try it at least.

> .I'll see how it goes building a site with the air-light to get better idea of css loading, example if @import is making another http request or build process writes @imports to css file:
> > Files prefixed by a _ (eg _blocks.css) are partials, compiled into their parent files using @import rules. PostCSS imports the contents of these partials during processing to become part of the parent file.

Ah, yes. We use SCSS and not CSS. This sounds like pretty much on how our sass/ directory is organized right now. We have partials as subfolders with _ prefixes (as in they won't compile into their own CSS files because they only contains parts of the project). 

SCSS imports actually do that compilation already, they combine imports to the file they are imported. See our [global.scss](https://github.com/digitoimistodude/air-light/blob/44499fc44c295deaae3f6bf5e4024f8f2348e837/sass/global.scss) for example which is combined and minified to production as [global.css](https://github.com/digitoimistodude/air-light/blob/44499fc44c295deaae3f6bf5e4024f8f2348e837/css/prod/global.css).

As for the wprig, I strongly dislike loading CSS via JS although I do know that CSS-in-JS is the way things are often done nowadays. Still, our sites built on Air load up from 20ms to 500ms depending on the site so I see little gain on trying to implement some sort of progressive CSS loading system.

We do have [postcss included](https://github.com/digitoimistodude/air-light/blob/44499fc44c295deaae3f6bf5e4024f8f2348e837/package.json#L34) in this theme but we mainly use some bits and pieces of it for tasks like stylelint and autoprefixer related things.

**Comment by LukaszJaro** [comment]
Yes thought you would appreciate another popular starter theme from a Nordic country :)

O awesome, so it's already being imported to one global CSS file with no extra http requests(although with HTTP/2 multiplexing the styles can be split into more css files to prevent downloading one big file)

By the way, wprig is not using JS to load the css. It uses wp_print_styles..example if you look at the top of index.php you'll see wp_rig()->print_styles( 'wp-rig-content' );

If you look at sidebar.php you'll see wp_rig()->print_styles( 'wp-rig-sidebar', 'wp-rig-widgets' );
If you look at front-page.php you'll see wp_rig()->print_styles( 'wp-rig-content', 'wp-rig-front-page' );

So you can see it's loading the styles based on that function, only when needed, if the module exists on the page.

HTTP/2 plus caching make this possible, which helps with render blocking CSS issues and reducing bandwidth.. it's actually based on this article about the future of loading CSS:

https://jakearchibald.com/2016/link-in-body/

wprig in this sense can be overkill.

**Comment by ronilaukkarinen** [comment]
> Yes thought you would appreciate another popular starter theme from a Nordic country :)

Of course! Amazing devs in scandinavia :)

> so it's already being imported to one global CSS file with no extra http requests(although with HTTP/2 multiplexing the styles can be split into more css files to prevent downloading one big file)

Indeed, this is where it's more of a matter of how you prefer and look at it. One could argue is it really better to have multiple CSS files instead of one - if that one is relatively small.

> If you look at sidebar.php you'll see wp_rig()->print_styles( 'wp-rig-sidebar', 'wp-rig-widgets' );
> If you look at front-page.php you'll see wp_rig()->print_styles( 'wp-rig-content', 'wp-rig-front-page' );

If we have - say -  a sidebar on front page, doesn't this mean we load up multiple css files on one page - hence multiple requests?

> So you can see it's loading the styles based on that function, only when needed, if the module exists on the page.

This gets really complicated if we also want to use Gutenberg and load up all the same stuff inside WordPress editor. It also gets overly hardcore to code on Network installs (the old term: "multisite"). Not to mention dynamic content, custom post types, search features, integrations, etc... then we use things like [swup.js](https://github.com/swup/swup) on some sites which is like SPA and known to relate to things loaded up first beforehand and then page switching etc. goes smoothly because everything is ALREADY loaded. See why even thinking of this may cause a little headache? :) 

> HTTP/2 plus caching make this possible, which helps with render blocking CSS issues and reducing bandwidth.. it's actually based on this article about the future of loading CSS

That was actually the article I read and there it says "The current state-of-the-art of loading CSS" where [loadCSS](https://github.com/filamentgroup/loadCSS) is advertised (which loads CSS via <script>. That article is from 2016 and the github repo is last updated a year ago so I wouldn't call it exactly "state of the art". 

I like this comparison [CSS tricks made in 2019](https://css-tricks.com/the-differing-perspectives-on-css-in-js/):

![Screen-Shot-2021-04-12-19-48-37 64](https://user-images.githubusercontent.com/1534150/114431381-1d2a2d80-9bc8-11eb-9de7-268b1cdc5f46.png)

If "more code" in our case means 50-200kb more, why bother?

But yeah, I like this type of pondering around the subject. Again, if you might know how to implement this, please do suggest changes in form of a PR or code snippets. I myself have been thinking this in the past but haven't yet found a really working solution for all types of situations.
